pkg/engine/constraints.go

```diff
@@ -0,0 +1,183 @@
+	}
+
+	// Edge is a struct that represents how we take in data about an edge in the resource graph
+	Edge struct {
+		Source core.ResourceId `yaml:"source"`
+		Target core.ResourceId `yaml:"target"`
+	}
+
+	// ApplicationConstraint is a struct that represents constraints that can be applied on the entire resource graph
+	ApplicationConstraint struct {
+		Operator ConstraintOperator `yaml:"operator"`
+		Node     core.ResourceId    `yaml:"node"`
+		Edge     Edge               `yaml:"edge"`
+	}
+
+	// ConstructConstraint is a struct that represents constraints that can be applied on a specific construct in the resource graph
+	ConstructConstraint struct {
+		Operator   ConstraintOperator `yaml:"operator"`
+		Target     core.ResourceId    `yaml:"target"`
+		Type       string             `yaml:"type"`
+		Attributes map[string]any     `yaml:"attributes"`
+	}
+
+	// EdgeConstraint is a struct that represents constraints that can be applied on a specific edge in the resource graph
+	EdgeConstraint struct {
+		Operator ConstraintOperator `yaml:"operator"`
+		Target   Edge               `yaml:"target"`
+		Node     core.ResourceId    `yaml:"node"`
+	}
+
+	// NodeConstraint is a struct that represents constraints that can be applied on a specific node in the resource graph.
+	// NodeConstraints are used to control intrinsic properties of a node in the resource graph
+	NodeConstraint struct {
+		BaseConstraint
+		Target   core.ResourceId
+		Property string
+		Value    any
+	}
+
+	// ConstraintScope is an enum that represents the different scopes that a constraint can be applied to
+	ConstraintScope string
+	// ConstraintOperator is an enum that represents the different operators that can be applied to a constraint
+	ConstraintOperator string
+)
+
+const (
+	ApplicationConstraintScope ConstraintScope = "application"
+	ConstructConstraintScope   ConstraintScope = "construct"
+	EdgeConstraintScope        ConstraintScope = "edge"
+	NodeConstraintScope        ConstraintScope = "node"
+
+	MustExistConstraintOperator      ConstraintOperator = "must_exist"
+	MustContainConstraintOperator    ConstraintOperator = "must_contain"
+	MustNotExistConstraintOperator   ConstraintOperator = "must_not_exist"
+	MustNotContainConstraintOperator ConstraintOperator = "must_not_contain"
+	AddConstraintOperator            ConstraintOperator = "add"
+	EqualsConstraintOperator         ConstraintOperator = "equals"
+)
+
+func (b *ApplicationConstraint) Scope() ConstraintScope {
+	return ApplicationConstraintScope
+}
+
+func (b *ConstructConstraint) Scope() ConstraintScope {
+	return ConstructConstraintScope
+}
+
+func (b *EdgeConstraint) Scope() ConstraintScope {
+	return EdgeConstraintScope
+}
+
+// DecodeYAMLNode is a helper function that decodes a yaml node into a struct representing different constraints
+func DecodeYAMLNode[T Constraint](node *yaml.Node) (constraint T, err error) {
+	constraint = reflect.New(reflect.TypeOf(constraint).Elem()).Interface().(T)
```

Comment:
```suggestion
func DecodeYAMLNode[T interface{ Constraint; *I }, I any](node *yaml.Node) (constraint T, err error) {
	constraint = new(I)
```
A little nicer to use generics instead of reflection.

Reply:
can go ahead and do that